ouroboros-consensus-test/test-consensus/Test/Consensus/Mempool.hs

```diff
@@ -77,6 +92,217 @@ tests = testGroup "Mempool"
+          curSize  = (msNumBytes . snapshotMempoolSize) snapshot
+          (res@(processed, unprocessed),_internalSt') =
+                     runTryAddTxs mpArgs internalSt (map fst txs)
+      in  counterexample ("Initial size: " <> show curSize)    $
+          classify (null processed)   "no transactions added"  $
+          classify (null unprocessed) "all transactions added" $
+          blindErrors res === expectedResult capacity curSize
+    where
+    -- | Convert 'MempoolAddTxResult' into a 'Bool':
+    -- isMempoolTxAdded -> True, isMempoolTxRejected -> False.
+    blindErrors
+      :: ([(GenTx TestBlock, MempoolAddTxResult blk)], [GenTx TestBlock])
+      -> ([(GenTx TestBlock, Bool)], [GenTx TestBlock])
+    blindErrors (processed, toAdd) = (processed', toAdd)
+      where
+        processed' = [ (tx, isMempoolTxAdded txAddRes)
+                     | (tx, txAddRes) <- processed ]
+
+    expectedResult
+      :: MempoolCapacityBytes
+      -> Word32  -- ^ Current mempool size
+      -> ([(GenTx TestBlock, Bool)], [GenTx TestBlock])
+    expectedResult (MempoolCapacityBytes capacity) = \curSize ->
+        go curSize [] txs
+      where
+        go
+          :: Word32
+          -> [(GenTx TestBlock, Bool)]
+          -> [(GenTx TestBlock, Bool)]
+          -> ([(GenTx TestBlock, Bool)], [GenTx TestBlock])
+        go curSize processed = \case
+          []
+            -> (reverse processed, [])
+          (tx, valid):txsToAdd'
+            | let curSize' = curSize + txSize tx
+            , curSize' <= capacity
+            -> go (if valid then curSize' else curSize)
+                  ((tx, valid):processed)
+                  txsToAdd'
+            | otherwise
+            -> (reverse processed, tx:map fst txsToAdd')
+
+{-------------------------------------------------------------------------------
+  Mempool Pure Test Infrastructure
+-------------------------------------------------------------------------------}
+
+data TestSetupSub = TestSetupSub
+  { tssSetup  :: TestSetup
+  , tssSubTxs :: [TestTx]
+  } deriving (Show)
+
+instance Arbitrary TestSetupSub where
+  arbitrary = do
+    testSetup <- arbitrary
+    subs      <- sublistOf (testInitialTxs testSetup)
+    return $ TestSetupSub testSetup subs
+
+  shrink TestSetupSub { tssSetup, tssSubTxs } =
+      [ TestSetupSub { tssSetup = setup', tssSubTxs }
+      | setup' <- shrink tssSetup
+      ]
+
+withInternalState
+  :: TestSetup
+  -> (MempoolArgs TestBlock -> InternalState TestBlock -> LedgerState TestBlock -> Property)
+  -> Property
+withInternalState testSetup@TestSetup {..} func =
+  let args               = MempoolArgs testLedgerConfig txSize testMempoolCapOverride
+      (slot, st')        = tickLedgerState testLedgerConfig (ForgeInUnknownSlot testLedgerState)
+      internalSt         = initInternalState testMempoolCapOverride zeroTicketNo slot st'
+      (_res,internalSt') = runTryAddTxs args internalSt testInitialTxs
```

Comment:
We expect all `testInitialTxs` to be valid, so I might be good to assert that none were invalid. When for some reason they are not all valid anymore, the tests will start failing, and debugging the failing tests will be much harder than seeing this assertion fail :slightly_smiling_face:.

Reply:
testInitialTxs comes with the comment: "These are all valid". I think nothing should break this assumption when running a test.